compiler/src/iree/compiler/Codegen/Common/BubbleUpOrdinalOps.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===- BubbleUpOrdinalOpPass.cpp -----------------------------------------===//
+//
+// The workgroup count computation when using slices needs the ordinal
+// annotation ops to be bubbled up as much as possible. This pass implements
+// patterns to bubble these operations up.
+//
+//===---------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+/// Replace the following sequence
+///
+/// ```mlir
+/// %1 = <cast> %0 : .. to index
+/// %2 = flow.dispatch.workload_ordinal %1, 0
+/// %3 = <some_op>(...%1)...
+/// ```
+///
+/// with
+///
+/// ```mlir
+/// %1 = <cast> %0 : .. to index
+/// %2 = flow.dispatch.workload_ordinal %1, 0
+/// %3 = <some_op>(...%2)...
+/// ```
+///
+/// to make all the uses flow through `flow.dispatch_workload_ordinal` ops.
+template <typename CastOpTy>
+struct BubbleUpAcrossCastOp
+    : public OpRewritePattern<IREE::Flow::DispatchWorkloadOrdinalOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::Flow::DispatchWorkloadOrdinalOp ordinalOp,
+                                PatternRewriter &rewriter) const override {
+    auto sourceCastOp = ordinalOp.getOperand().getDefiningOp<CastOpTy>();
+    if (!sourceCastOp || sourceCastOp->hasOneUse()) {
+      return failure();
+    }
```

Comment:
Maybe we can use `CastOpInterface`, so it can be applied to all the ops that implements the interface? In this context, we don't need template.

And we can add a check about if the target element type is IndexType or not.

Reply:
I am trying to slow walk this... if it fails, i want it to fail explicitly so that I can make sure things are working as expected.